File: 2021/day05_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any

type Coords = tuple[int, int]
type Vector = tuple[Coords, Coords]


def solution(content: list[str]) -> Any:
    straight = [vec for vec in [parse(line) for line in content] if is_straight(vec)]
    max_x, max_y = get_limit(straight)
    return len([coord for coord in each(max_x, max_y) if find_two(coord, straight)])


def find_one(coords: Coords, vecs: list[Vector]):
    for vec in vecs:
        if is_in_line(coords, vec):
            return True
    return False


def find_two(coords: Coords, vecs: list[Vector]):
    for index, vec in enumerate(vecs):
        if is_in_line(coords, vec):
            return find_one(coords, vecs[index + 1 :])
    return False


def is_in_line(coords: Coords, vec: Vector):
    x, y = coords
    ((sx, sy), (ex, ey)) = vec
    if x == sx == ex:
        return sy <= y <= ey or ey <= y <= sy
    if y == sy == ey:
        return sx <= x <= ex or ex <= x <= sx
    return False


def each(mx: int, my: int):
    for x in range(mx + 1):
        for y in range(my + 1):
            yield (x, y)


def get_limit(vecs: list[Vector]) -> Coords:
    x, y = 0, 0
    for (sx, sy), (ex, ey) in vecs:
        x = max(x, sx, ex)
        y = max(y, sy, ey)
    return (x, y)


def is_straight(vec: Vector):
    start, end = vec
    sx, sy = start
    ex, ey = end
    return sx == ex or sy == ey


def parse(line: str) -> Vector:
    match = re.match(r"(\d+),(\d+) -> (\d+),(\d+)", line)
    if match is None:
        raise ValueError(f"Invalid line {line}")
    return ((int(match[1]), int(match[2])), (int(match[3]), int(match[4])))


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2021/day05-1.ts
type Coords = {
  x: number;
  y: number;
};

type Vector = {
  start: Coords;
  end: Coords;
};

function solution(content: string) {
  const straight = content.split("\n").map(parse).filter(isStraight);
  const { x: mx, y: my } = getLimit(straight);
  return each(mx, my).filter((cc) => findTwo(cc, straight)).length;
}

function findOne(coords: Coords, vecs: Vector[]) {
  return !!vecs.find((v) => isInLine(coords, v));
}

function findTwo(coords: Coords, vecs: Vector[]) {
  const index = vecs.findIndex((v) => isInLine(coords, v));
  if (index === -1) return false;
  return findOne(coords, vecs.slice(index + 1));
}

function isInLine(
  { x, y }: Coords,
  { start: { x: sx, y: sy }, end: { x: ex, y: ey } }: Vector
) {
  if (allEqual(x, sx, ex)) return isBetween(y, sy, ey);
  if (allEqual(y, sy, ey)) return isBetween(x, sx, ex);
  return false;
}

function allEqual(x: number, y: number, z: number) {
  return x === y && y === z;
}

function isBetween(x: number, left: number, right: number) {
  return (left <= x && x <= right) || (right <= x && x <= left);
}

function each(mx: number, my: number) {
  return Array.from({ length: mx + 1 })
    .map((_, x) => Array.from({ length: my + 1 }).map((_, y) => ({ x, y })))
    .flat();
}

function getLimit(vecs: Vector[]): Coords {
  return vecs.reduce(
    (acc, cur) => ({
      x: Math.max(acc.x, cur.start.x, cur.end.x),
      y: Math.max(acc.y, cur.start.y, cur.end.y),
    }),
    { x: 0, y: 0 } as Coords
  );
}

function isStraight({
  start: { x: sx, y: sy },
  end: { x: ex, y: ey },
}: Vector) {
  return sx === ex || sy === ey;
}

function parse(line: string): Vector {
  const match = line.match(/(\d+),(\d+) -> (\d+),(\d+)/);
  if (match === null) throw new Error(`Invalid line ${line}`);
  return {
    start: { x: Number(match[1]), y: Number(match[2]) },
    end: { x: Number(match[3]), y: Number(match[4]) },
  };
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
